### DISCUSSION TITLE
PPL vs PAD for IO

### DISCUSSION THREAD

**Discussion by lustefan97** [The-OpenROAD-Project/OpenROAD#4155] [Q&A]
Hello,

I have a question regarding the differences between the PPL and PAD tools, especially the PAD bump placement for flip chip designs. I don't really get the differences between the PPL and PAD tools when it comes to IO placed on the top metal layer, and their "relationship" with the GPL tool.

I am right now doing some tests using different designs using the PPL tool. It is used with GPL first by placing the IO without considering GPL, then doing Global PLacement followed by a Pin PLacement after it, finishing by a final GPL. I read that it was due to some circular dependency between the two tools, so this was the way things were done. 
Now I am using PPL to do some 3D Face-to-Face implementation, and am using the `define_pin_shape_pattern` command to place the IO on the top metal layer.

From what I understood PPL doesn't have/need a LEF object for the bump as it will create a metal shape in place of the pin, based off of the constraint given to the tool by the user before being routed in the subsequent steps.

I am now trying to understand what the PAD tool really is about, how it differs from what PPL is and what his use case may be, but I am a bit confused when it comes to 3D interconnect.

I understand that PAD is used for chip level connection, which raise a question for me (that is asked below), that requires a LEF object for top layer bumps.
When it comes to flip chip design, the PAD tool seems to be the default candidate but something is bothering me,  it is simply that there is no pin position optimization in it as opposed to PPL (which would probably not give a minimal "3D" wire length), that the bump assignement is done manually, and that in the example script given in _OpenROAD/src/pad/test_ link the bumps to an IO ring which I don't think is the goal in my application.

About the pin position optimization, I was told by @rovinski in #4061 that GPL will still do some HPWL optimization (and that the team would probably be interested in implementing 3D interconnect as part of the PPL tool), but at first sight it doesn't seem to be the most optimal way to proceed for "3D" routing since the assignement will be done manually. This leads me to the second question about assignment : Since I will probably need to have thousands of IOs on the top layer, manual assignment seems tedious if not even suboptimal due to the lack of automatic bump repositioning. 
I get that using a tool for flip-chip design would be make more sense but I must say that, when it comes to top layer bumps, I don't see really the interest in going for PAD rather than PPL

My questions thus are : 

0) What are the exact mechanisms that OpenROAD use to minimize wire length w.r.t. IO pins ?
1) If PAD is used for chip level connections then what is the use case for PPL ?
2) What are the inherent differences between PAD and PPL ?
3) If PAD is the default for flip-chip design, why implementing such a thing as part of PPL ?
4) Does will OpenROAD route differently when using PAD or PPL ? Especially when talking about memory usage ?
5) Will the use of a LEF object in the PAD tool create a difference with regard to the pin created by PPL ? Could I use a LEF object for PPL instead to benefit from the placement optimization?
6) Could we imaging eventually a flow mixing PAD and PPL, where the PPL tool is used for 3D interconnections ?

**Comment by maliberty** [comment]
What is driving the bumps?   Usually you need an IO driver and that means you need to use pad.

**Reply by rovinski** [reply]
It depends on the IO wirelength. There some 3D technologies that are close enough where you can actually get away with a max size buffer. But generally, yes, an IO driver is needed.

**Reply by maliberty** [reply]
If we are dealing with 3d uBumps then that is a different story - hence the question.

**Reply by rovinski** [reply]
A couple notes:
* Honestly I did not know about `define_pin_shape_pattern` before. For your use case, it seems like the right approach along with PPL.
* With respect to the I/O driver, an easy first-order check is to derive the expected cap of the 3D interconnect, set the load on the pin, and see if it passes slew checks in STA. If it fails horribly, then that's a good indication you will need a custom I/O driver (either a pad cell or your own). Even if it doesn't fail, you would probably have to consider electromigration on the output buffer.
* If you do find you need a custom I/O driver, there is not a built-in mechanism to handle that right now. You would need to incorporate an area I/O cell of some kind, or make a custom I/O driver cell that suits your use case.
* Depending on what you are trying to model, you may not even care about modeling the I/O driver yet and can simply ignore it for now. But you should definitely be aware of any modeling shortcomings.

**Comment by QuantamHD** [comment]
I think the biggest difference between pad and ppl is that pad tries to deal with io cells. Typically the foundry will require you to place specialized io cells that deal with esd and what not.

ppl does a similar pin placement, but doesn't deal with io cells. If you don't need io cells between your dies then ppl probably is the solution you should use or enhance

**Reply by lustefan97** [reply]
@QuantamHD Thanks for your answer ! Would you then confirm that, indeed, a metal shape is created at the place where the IO pad will be created when using PPL, in the same fashion that the PAD tool will create a pad location based off of the LEF file provided ?

Also, is the routing handled by the same tool and in the same fashion whether it is with PAD or PPL ? By that I mean, will there any difference in results, method, or CPU/memory usage when using PAD vs PPL ?

**Reply by maliberty** [reply]
They are completely different methods, different code, and will give different results.  Pad frames are a particular style of design that doesn't resemble regular placement.

**Reply by lustefan97** [reply]
But when it comes to routing, TritonFast and TritonRoute are still used ? Or will OpenROAD use another method depending on if it routes a pad from PAD or a pin from PPL ?

**Reply by maliberty** [reply]
There is no TritonFast - I suppose you mean FastRoute.  With pad it has its own RDL router as that is the usual use case for pads and bumps.

**Reply by lustefan97** [reply]
Yes I meant FastRoute sorry. I do have one more question, which relates to the use case differences between PAD and PPL. I get that PAD is used for chip-level connection, but then what would PPL be used for ? 

Would it be used to create functional blocks that would interface themselves with other blocks using the metal pads created ? For instance I have in mind the Ibex core examples, where its IOs are placed using PPL on the periphery of the chip, it has no IO ring or bump.

**Reply by maliberty** [reply]
Yes ppl is mostly for block pin placement.

**Reply by lustefan97** [reply]
Ok, I understand better now ! Thanks for your help

**Reply by lustefan97** [reply]
I do have one more question regarding PAD and more specifically the routing tool. 

I see that the `rdl_route` command ask for only one layer, and in the example scripts in _OpenROAD/src/pad/test_ the bump seem to be connected to the IO cells using the layer specified (with the bumps being routed from the bottom if the layer is not the top-most layer). 

My questions are simply : 

1) Does this tool expect principally to route bumps to IO cells ?
2) There is a [-net] option to this command, which leads me to believe that I could use it to route a buffer of standard cell to a bump, but then wouldn't the fact that only one layer is available for routing be an issue if I have hundreds or even thousands of bumps to route to standard cells ? Especially considering the fact that the only wire length optimization would come from the placement tools ?

**Reply by maliberty** [reply]
The RDL router is for connecting IO pads to bumps exclusively on the RDL layer using pattern routing.  It is not for use with standard cells.
